### PR TITLE
*: Add support for GET_LOCK() / RELEASE_LOCK() / RELEASE_ALL_LOCKS()

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/14994

Problem Summary:

This supports the `GET_LOCK()` / `RELEASE_LOCK()`  / `RELEASE_ALL_LOCKS()` functions previously proposed in the design document @ https://github.com/pingcap/tidb/pull/33406

### What is changed and how it works?

Functions `GET_LOCK`, `RELEASE_LOCK`, `RELEASE_ALL_LOCKS` are no longer noops. The functionality is now implemented.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
Deployed 2x tidb, 1x tidb/pd. Confirmed functionality works correctly.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
Changes the behavior of `tidb_enable_noop_funcs`, so documentation needs updating.
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
TiDB now supports the GET_LOCK() / RELEASE_LOCK() / RELEASE_ALL_LOCKS() advisory lock functions.
```

### CODE REVIEW

**Comment by yahonda** [comment]
Let me confirm if my understanding is correct that based on this implementation, the maximum value of advisory lock timeout only depends on `innodb_lock_wait_timeout` whose max value is `1073741824` sec, about 34 years.
And it should not be limited by the max value of `max-txn-ttl` timeout 1 hour.

**Comment by yahonda** [comment]
It looks like the get_lock timeout is limited by 1 hour.


1. Get lock with second argument as 3800 sec.

```sql
mysql> select get_lock('foo', 3800);
+-----------------------+
| get_lock('foo', 3800) |
+-----------------------+
|                     1 |
+-----------------------+
1 row in set, 1 warning (0.00 sec)
```

2. Attempt to get the same lock, which results this failure

```sql
mysql> select get_lock('FOO', -1);
+---------------------+
| get_lock('FOO', -1) |
+---------------------+
|                   0 |
+---------------------+
1 row in set, 1 warning (1 hour 0.01 sec)
mysql> show warnings;
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                            |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect get_lock value: '%!s(int64=-00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001)' |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql>
```

- ./bin/tidb-server STDOUT

```
[2022/04/14 16:37:56.632 +09:00] [INFO] [2pc.go:1092] ["ttlManager live up to its lifetime"] [txnStartTS=432516176655155200] [uptime=3600207] [maxTxnTTL=3600000]
[2022/04/14 16:38:01.229 +09:00] [WARN] [session.go:1946] ["run statement failed"] [schemaVersion=34] [error="[tikv:1205]Lock wait timeout exceeded; try restarting transaction"] [session="{\n  \"currDBName\": \"\",\n  \"id\": 0,\n  \"status\": 3,\n  \"strictMode\": true,\n  \"txn\": \"432516177913446400\",\n  \"user\": null\n}"]
```

**Comment by morgo** [comment]
> Let me confirm if my understanding is correct that based on this implementation, the maximum value of advisory lock timeout only depends on `innodb_lock_wait_timeout` whose max value is `1073741824` sec, about 34 years. And it should not be limited by the max value of `max-txn-ttl` timeout 1 hour.

The maximum for `innodb_lock_wait_timeout` was changed to 1hr to match the default of `max-txn-ttl`. See: https://github.com/pingcap/tidb/pull/33473

For this:
```sql
mysql> show warnings;
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                            |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect get_lock value: '%!s(int64=-00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001)' |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

I'll fix the message. But the warning is intentional - it's not a MySQL behavior, but since we can't do unlimited I thought this was important.

File: expression/builtin_miscellaneous.go [hunk 1]
```diff
@@ -221,10 +271,27 @@ func (b *builtinReleaseLockSig) Clone() builtinFunc {
...
+		return 0, false, errUserLockWrongName.GenWithStackByArgs("NULL")
+	}
+	if lockName == "" || len(lockName) > 64 {
+		return 0, false, errUserLockWrongName.GenWithStackByArgs(lockName)
+	}
+
+	// Lock names are case insensitive. Because we can't rely on collations
+	// being enabled on the internal table, we have to lower it.
+	lockName = strings.ToLower(lockName)
+	released := int64(0)
```
Comment (mjonss) [inline expression/builtin_miscellaneous.go:290]: In GET_LOCK() you also check if the lower case lockName has len <= 64, should that be done here as well?

File: expression/function_traits.go [hunk 2]
```diff
@@ -226,13 +226,9 @@ var mutableEffectsFunctions = map[string]struct{}{
...
 
-// some functions like "get_lock" and "release_lock" currently do NOT have
-// right implementations, but may have noop ones(like with any inputs, always return 1)
+// some functions do NOT have right implementations, but may have noop ones(like with any inputs, always return 1)
 // if apps really need these "funcs" to run, we offer sys var(tidb_enable_noop_functions) to enable noop usage
-var noopFuncs = map[string]struct{}{
-	ast.GetLock:     {},
-	ast.ReleaseLock: {},
-}
+var noopFuncs = map[string]struct{}{}
```
Comment (mjonss) [inline expression/function_traits.go:231]: Are there any other noop functions that may use this in the future? Or should we remove this?

File: expression/integration_test.go [hunk 3]
```diff
@@ -249,6 +249,122 @@ func TestBuiltinFuncJsonPretty(t *testing.T) {
...
+	tk.MustQuery("SELECT release_lock('AbC')").Check(testkit.Rows("1"))
+
+	// Release unacquired LOCK and previously released lock
+	tk.MustQuery("SELECT release_lock('randombytes')").Check(testkit.Rows("0"))
+	tk.MustQuery("SELECT release_lock('abc')").Check(testkit.Rows("0"))
+
+	// GetLock with integer name, 64, character name.
+	_ = tk.MustQuery("SELECT get_lock(1234, 10)")
+	_ = tk.MustQuery("SELECT get_lock(REPEAT('a', 64), 10)")
+	_ = tk.MustQuery("SELECT release_lock(1234), release_lock(REPEAT('aa', 32))")
```
Comment (mjonss) [inline expression/integration_test.go:312]: haha, you tricked me here :) Any reason for the difference of 'a' x 64 vs 'aa' x 32? Why not check the results as well?

File: expression/integration_test.go [hunk 4]
```diff
@@ -249,6 +249,122 @@ func TestBuiltinFuncJsonPretty(t *testing.T) {
...
+	tk.MustQuery("SELECT release_lock('mygloballock')").Check(testkit.Rows("0"))
+
+	// Test someone else has the lock with short timeout.
+	tk2 := testkit.NewTestKit(t, store)
+	tk2.MustQuery("SELECT get_lock('mygloballock', 1)").Check(testkit.Rows("1"))
+	tk.MustQuery("SELECT get_lock('mygloballock', 1)").Check(testkit.Rows("0"))  // someone else has the lock
+	tk.MustQuery("SELECT release_lock('mygloballock')").Check(testkit.Rows("0")) // never had the lock
+	// try again
+	tk.MustQuery("SELECT get_lock('mygloballock', 0)").Check(testkit.Rows("0"))  // someone else has the lock
+	tk.MustQuery("SELECT release_lock('mygloballock')").Check(testkit.Rows("0")) // never had the lock
```
Comment (mjonss) [inline expression/integration_test.go:363]: Could not this be used for implementing IS_FREE_LOCK()?
I.e. first check if the session already have the lock, if not do a get_lock with 0 timeout and if we got the lock immediately release it?

File: expression/integration_test.go [hunk 5]
```diff
@@ -330,6 +445,8 @@ func TestMiscellaneousBuiltin(t *testing.T) {
 	result.Check(testkit.Rows("1"))
 	result = tk.MustQuery(`SELECT RELEASE_LOCK('test_lock1');`)
 	result.Check(testkit.Rows("1"))
+	result = tk.MustQuery(`SELECT RELEASE_LOCK('test_lock3');`) // not acquired
+	result.Check(testkit.Rows("0"))
 }
```
Comment (mjonss) [inline expression/integration_test.go:450]: For completeness only, maybe also add RELEASE_ALL_LOCKS?

File: session/session.go [hunk 6]
```diff
@@ -2544,6 +2607,9 @@ func (s *session) Close() {
 			logutil.BgLogger().Error("release table lock failed", zap.Uint64("conn", s.sessionVars.ConnectionID))
 		}
 	}
+	if s.hasAdvisoryLocks() {
```
Comment (mjonss) [inline session/session.go:2610]: Is `hasAdvisoryLocks()` needed, can we not just call s.ReleaseAllAdvisoryLocks() regardless?

**Review by mjonss — APPROVED** [review]
When a TiDB server with advisory locks (or any pessimistic transaction) dies (kill -9 for example) how are those locks handled? I did not find any documentation on this...

Mostly questions and still looks good to me, so approved, feel free to consider my suggestions.

File: expression/function_traits.go (hunk 2 above)
Comment (morgo) [reply expression/function_traits.go:231]: Yes, I remember when I wrote this a couple of weeks ago that there was a function that we should add to this list. But I can't remember it right now. We can remove it in a future refactor if that's not the case.

File: expression/integration_test.go (hunk 3 above)
Comment (morgo) [reply expression/integration_test.go:312]: Yes, I did do this intentionally. I thought as well as testing other data types (int: 1234) I should test expressions which should resolve to strings, and expressions that match.

I will add result checking too.

File: expression/integration_test.go (hunk 4 above)
Comment (morgo) [reply expression/integration_test.go:363]: Yes, I guess that's possible.

But it's not the most elegant. If I learn the transaction code, I assume I could implement this correctly by checking in TiKV if the lock is held. So I will try and keep out of scope for now.

**Comment by morgo** [comment]
> When a TiDB server with advisory locks (or any pessimistic transaction) dies (kill -9 for example) how are those locks handled? I did not find any documentation on this...

For clean shutdown and kill, I've verified it works. For unsafe shutdown it could take up to the 1hr timeout to clear, I'm not sure if there are mechanisms in place which could shorten it.

**Comment by morgo** [comment]
@mjonss thanks for the review. PTAL again :-)

File: expression/integration_test.go (hunk 4 above)
Comment (morgo) [reply expression/integration_test.go:373]: I thought about this a little more. I think you are right: some implementation is better than no implementation. We know `IS_FREE_LOCK()` is used in the rails test suite (but not in rails itself).

I will plan to add it for this sprint, but in a separate PR as this one is already getting large.

**Comment by yahonda** [comment]
> The maximum for innodb_lock_wait_timeout was changed to 1hr to match the default of max-txn-ttl. See: https://github.com/pingcap/tidb/pull/33473

Understood the limitation.

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**